### PR TITLE
Commit sub epoch segment/fix cache

### PR DESCRIPTION
**Pull request by almogdepaz**
add explicit commit at end of persist_sub_epoch_challenge_segments
make sure all segments are in db on restart
add segment to db in short sync

### CODE REVIEW

**Comment by almogdepaz** [comment]
reverted the lock, it caused a deadlock because we are saving segments when we are syncing (under blockchain lock)
added code for adding of segments to cache on gossip

File: chia/full_node/full_node.py [hunk 1]
```diff
@@ -1003,6 +1003,9 @@ async def respond_block(
         if peak.height % 1000 == 0 and not self.sync_store.get_sync_mode():
             await self.sync_store.clear_sync_info()  # Occasionally clear sync peer info
         self._state_changed("block")
+        record = self.blockchain.block_record(block.header_hash)
```
Comment (mariano54) [inline chia/full_node/full_node.py:1006]: This will slow down the node, right? Could we do it every sub-epoch instead of every block?

Comment (almogdepaz) [reply chia/full_node/full_node.py:1006]: it happenes only when sub epoch ends though

Comment (almogdepaz) [reply chia/full_node/full_node.py:1006]: the block_record call is a call to the cache

Comment (mariano54) [reply chia/full_node/full_node.py:1006]: I meant the next line, which creates the segments

Comment (almogdepaz) [reply chia/full_node/full_node.py:1006]: that happens only at the end of a sub epoch

Comment (almogdepaz) [reply chia/full_node/full_node.py:1006]: @mariano54 i can move it to the end of short_sync_backtrack that might be a better place to call it

Comment (mariano54) [reply chia/full_node/full_node.py:1007]: Oh yeah I can't believe I missed that, you're right.

**Review by mariano54 — DISMISSED** [review]

File: chia/full_node/full_node.py (hunk 1 above)
Comment (mariano54) [reply chia/full_node/full_node.py:1007]: No need to move it into backtrack. That still get's called once per block, so it's the same.

Comment (almogdepaz) [reply chia/full_node/full_node.py:1007]: maybe we should do some better concurrency here but im not sure how to do that without slowing down things under 
blockchain.lock

Comment (almogdepaz) [reply chia/full_node/full_node.py:1007]: i can probably clean all the requests if we go in to long sync, i think that should work without slowing down gossip

Comment (almogdepaz) [reply chia/full_node/full_node.py:1007]: it is probably isn't that bad, it will just overwrite but it will use the correct segment just redundant work

Comment (almogdepaz) [reply chia/full_node/full_node.py:1014]: @mariano54 i added tracking for the task, and cancelling if they stay for the next sub_epoch/long sync

**Comment by almogdepaz** [comment]
closing moving to clean branch

**Pull request closed without merging**